Gems/EditorPythonBindings/Code/Source/PythonSystemComponent.cpp

```diff
@@ -258,6 +258,85 @@ namespace EditorPythonBindings
         void Finalize() override {}
     };
 
+    // Manages the acquisition and release of the Python GIL (Global Interpreter Lock).
+    // Used by PythonSystemComponent to lock the GIL when executing python.
+    class PythonSystemComponent::PythonGILScopedLock final
+    {
+    public:
+        PythonGILScopedLock(AZStd::recursive_mutex& lock, int& lockRecursiveCounter, bool tryLock = false);
+        ~PythonGILScopedLock();
+
+        bool IsLocked() const;
+
+    protected:
+        void Lock(bool tryLock);
```

Comment:
>Why combine the underlying lock and try_lock functionality into a single function like this?
I had the logic of `tryLock` in `PythonGILScopedLock` constructor before, but then that meant `m_lock.try_lock()` and `m_lock.lock()` calls happened in the constructor, but it's actually nicer to have them inside the `Lock` function, in the same way that `Unlock` function has the `m_lock.unlock()`. Notice that `Lock` is protected, to served the constructor, not meant to be called by the user.

> It also discards the functionality of try_lock returning a boolean
This class has to be used in an scoped manner, the call comes from the constructor, so it would never have returned the boolean to the user anyway.

> Is the user intended to call IsLocked() to evaluate whether the tryLock variant succeeded?
Correct, you can see how it's used in `PythonSystemComponent::TryExecuteWithLock` function. But `IsLocked` does not only work when `tryLock` is true, it also works when `tryLock` is false, it'll always return true, but the function still makes sense.